src/main/java/com/xebialabs/overthere/smb/SmbFile.java

```diff
@@ -0,0 +1,359 @@
+                    return wrapped.read();
+                }
+
+                @Override
+                public int read(byte[] b) throws IOException {
+                    return wrapped.read(b);
+                }
+
+                @Override
+                public int read(byte[] b, int off, int len) throws IOException {
+                    return wrapped.read(b, off, len);
+                }
+
+                @Override
+                public long skip(long n) throws IOException {
+                    return wrapped.skip(n);
+                }
+
+                @Override
+                public void close() throws IOException {
+                    logger.debug("Closing SMB input stream for {}", getSharePath());
+                    wrapped.close();
+                    getShare().close(file.getFileId());
+                }
+            });
+        } catch (TransportException e) {
+            throw new RuntimeIOException(format("Cannot open %s for reading: %s", getSharePath(), e.toString()), e);
+        }
+    }
+
+    @Override
+    public OutputStream getOutputStream() {
+        logger.debug("Opening SMB output stream for {}", getSharePath());
+        try {
+            SMB2CreateDisposition createDisposition = SMB2CreateDisposition.FILE_OVERWRITE_IF;
+            if (!overwrite) createDisposition = SMB2CreateDisposition.FILE_CREATE;
+            final File file = getShare().openFile(getPathOnShare(),
+                    EnumSet.of(AccessMask.GENERIC_WRITE), createDisposition);
+
+            final OutputStream wrapped = file.getOutputStream();
+
+            return asBuffered(new OutputStream() {
+
+                @Override
+                public void write(int b) throws IOException {
+                    wrapped.write(b);
+                }
+
+                @Override
+                public void write(byte[] b, int off, int len) throws IOException {
+                    wrapped.write(b, off, len);
+                }
+
+                @Override
+                public void write(byte[] b) throws IOException {
+                    wrapped.write(b);
+                }
+
+                @Override
+                public void flush() throws IOException {
+                    wrapped.flush();
+                }
+
+                @Override
+                public void close() throws IOException {
+                    logger.debug("Closing SMB output stream for {}", getSharePath());
+                    wrapped.close();
+                    getShare().close(file.getFileId());
+                }
+            });
+        } catch (TransportException e) {
+            throw new RuntimeIOException(format("Cannot open %s for writing: %s", getSharePath(), e.toString()), e);
+        }
+    }
+
+    @Override
+    public void setExecutable(boolean executable) {
+        // the execute permission does not exist on Windows
+    }
+
+    @Override
+    public void delete() {
+        String sharePath = getPathOnShare();
```

Comment:
extract a private method `delete(boolean recursive)` and reuse that with the `deleteRecursive` method below. The implementation is too similar not to.